--- org/GeoIP.c	Sun Nov 25 05:27:25 2007
+++ GeoIP.c	Sun Sep 14 11:23:00 2008
@@ -1,3 +1,7 @@
+#define mmap(a,b,c,d,e,f) 0
+#define munmap(a,b) 0
+#define MAP_FAILED 0
+
 /* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 2; tab-width: 2 -*- */
 /* GeoIP.c
  *
@@ -521,7 +525,7 @@
 	GeoIP * gi;
 	size_t len;
 
-#ifdef WIN32
+#if 0
 	WSADATA wsa;
 	if (WSAStartup(MAKEWORD(1, 1), &wsa) != 0)
 		return NULL;
